core/coreunix/add.go

```diff
@@ -22,8 +22,12 @@ import (
 
 var log = eventlog.Logger("coreunix")
 
-// Add builds a merkledag from the a reader, pinning all objects to the local
-// datastore. Returns a key representing the root node.
+// DEPRECATED: Add builds a merkledag from the a reader, pinning all
+// objects to the local datastore. Returns a key representing the root
+// node.
+//
+// This function is deprecated and will be removed in 0.4.0.  You
+// should use shell/fsnode's AddFromReader instead.
```

Comment:
On Wed, Apr 29, 2015 at 01:41:12PM -0700, Juan Batiz-Benet wrote:
> I'm not 100% sure about this right now, but it seems to me that all
> the things that the shell can do, will likely be in core as
> well. (the shell is mostly just a wrapper). Perhaps I should spend
> time charting out what shell and core should look like, i can do
> that next sprint.

I'd like to shift importer.BuildDagFromReader into the core (as
core/unixfs's AddFromReader?  In any case it should match what's
currently my shell/fsnode AddFromReader, with shell → core).  That
seems like it has all of our current bells and whistles for creating a
DAG node from a reader, so it's a good low-level interface.

> the core is where the bulk of the operations will remain, the shell
> is just a very simplified interface that makes it easier for users
> to use ipfs nodes that are either embedded or over an rpc transport.

Exactly, I just think the function known as BuildDagFromReader *is*
the low-level interface (although I don't mind if you want to shuffle
its signature around when we move it into core).  If someone wants the
high-level interface, they can use shell/unixfs (or shell/fsnode, or
whatever we call it).  But I see no reason for a complete copy of the
high-level API into the low-level core space.

> okay, i haven't made things very clear with all this. we should spec
> out both the shell and the core should be -- on their own as
> interface specs -- before deciding what to deprecate, or actually
> deprecating, from the current interfaces. trying to decide one thing
> goes here and not there without the full picture -- and with clearly
> different mental models -- is going to cause undue decision
> conflicts. i'll make it a priority to get that out next week.

I'm fine with this, but I'm also fine deciding now that the current
coreunix.Reader is *not* going to be an appropriate low-level
interface (it's missing at least a knob to select your desired
chunker).  Anyhow, we can certainly put off the deprecation (and maybe
this whole PR?) until you've had time to think this over, but I think
it will be easier to evolve towards a stable API spec than it will be
to design it all at once before focusing on implementation.